client/lease.py

```diff
@@ -0,0 +1,129 @@
+from api.xline.rpc_pb2 import (
+    LeaseGrantRequest,
+    LeaseGrantResponse,
+    LeaseRevokeRequest,
+    LeaseRevokeResponse,
+    LeaseKeepAliveRequest,
+    LeaseTimeToLiveRequest,
+    LeaseTimeToLiveResponse,
+    LeaseLeasesRequest,
+    LeaseLeasesResponse,
+)
+
+
+class LeaseIdGenerator:
+    """
+    Generator of unique lease id
+    Note that this Lease Id generation method may cause collisions,
+    the client should retry after informed by the server.
+
+    Attributes:
+        id: The current lease id.
+    """
+
+    lease_id: int
+
+    def __init__(self) -> None:
+        self.lease_id = int.from_bytes(random.randbytes(8), "big")
+
+    def next(self) -> int:
+        """Generate a new `leaseId`."""
+        lease_id = self.lease_id
+        self.lease_id += 1
+        if lease_id == 0:
+            return self.next()
+        return lease_id & 0x7FFF_FFFF_FFFF_FFFF
+
+
+class LeaseClient:
+    """
+    Client for Lease operations.
+
+    Attributes:
+        name: Name of the LeaseClient, which will be used in CURP propose id generation.
+        curp_client: The client running the CURP protocol, communicate with all servers.
+        lease_client: The lease RPC client, only communicate with one server at a time.
+        token: The auth token.
+        id_gen: Lease Id generator.
+    """
+
+    name: str
+    curp_client: CurpClient
+    lease_client: LeaseStub
+    token: Optional[str]
+    id_gen: LeaseIdGenerator
+
+    def __init__(
+        self, name: str, curp_client: CurpClient, channel: Channel, token: Optional[str], id_gen: LeaseIdGenerator
+    ) -> None:
+        self.name = name
+        self.curp_client = curp_client
+        self.lease_client = LeaseStub(channel=channel)
+        self.token = token
+        self.id_gen = id_gen
+
+    async def grant(self, req: LeaseGrantRequest) -> LeaseGrantResponse:
+        """
+        Creates a lease which expires if the server does not receive a keepAlive
+        within a given time to live period. All keys attached to the lease will be expired and
+        deleted if the lease expires. Each expired key generates a delete event in the event history.
+        """
+        if req.ID == 0:
+            req.ID = self.id_gen.next()
+        request_with_token = RequestWithToken(token=self.token, lease_grant_request=req)
+        propose_id = self.generate_propose_id()
+        cmd = Command(
+            request=request_with_token,
+            propose_id=propose_id,
+        )
+        er, _ = await self.curp_client.propose(cmd, True)
+        return er.lease_grant_response
+
+    async def revoke(self, req: LeaseRevokeRequest) -> LeaseRevokeResponse:
+        """
+        Revokes a lease. All keys attached to the lease will expire and be deleted.
+        """
+        res: LeaseRevokeResponse = await self.lease_client.LeaseRevoke(req)
+        return res
+
+    async def keep_alive(self, req: Generator[LeaseKeepAliveRequest, None, None]):
```

Comment:
Maybe make this more user friendly? The users needs to pass in a `Gernerator` everytime they want to call keep alive.

Reply:
Done.